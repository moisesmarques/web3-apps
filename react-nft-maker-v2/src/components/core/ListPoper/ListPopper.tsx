import React, { useState, useEffect, useRef } from 'react';

import ClickAwayListener from '@mui/base/ClickAwayListener';
import Paper from '@mui/material/Paper';
import Popper, { PopperPlacementType } from '@mui/material/Popper';
import cn from 'classnames';

import ListPopperStyled from './ListPopper.styles';

interface Props {
  Button: React.ReactElement;
  List: React.ReactElement;
  open: boolean;
  onClose: () => void;
  popperClassName?: string;
  placement?: PopperPlacementType;
}

export const ListPopper: React.FC<Props> = ({
  Button,
  List,
  open,
  onClose,
  popperClassName,
  placement = 'bottom-end',
}) => {
  const [isFirstRender, setIsFirstRender] = useState(true);
  // on first render there's no ref created which is required by Popper
  // this workaround forces rerender to make the ref accessible by Popper component
  useEffect(() => {
    if (isFirstRender) {
      setIsFirstRender(false);
    }
  }, [isFirstRender]);

  const anchorRef = useRef<HTMLDivElement>(null);

  const handleCloseList = (event: MouseEvent | TouchEvent) => {
    if (anchorRef.current && anchorRef.current.contains(event.target as HTMLElement)) {
      return;
    }
    onClose();
  };

  return (
    <ListPopperStyled>
      <div ref={anchorRef}>{Button}</div>
      <Popper
        open={open && !isFirstRender}
        placement={placement}
        anchorEl={anchorRef.current}
        className={cn('dropDown', popperClassName)}
      >
        <ClickAwayListener onClickAway={handleCloseList}>
          {/* ClickAwayListener children should be enwrapped with any component to make it work */}
          <Paper elevation={1}>{List}</Paper>
        </ClickAwayListener>
      </Popper>
    </ListPopperStyled>
  );
};

export default ListPopper;
