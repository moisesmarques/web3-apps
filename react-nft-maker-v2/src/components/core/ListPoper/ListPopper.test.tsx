import React from 'react';

import { render, fireEvent } from '@testing-library/react';

import ListPopper from './ListPopper';

const props = {
  Button: <button type="button">Open list</button>,
  List: <div>List of lists</div>,
  onClose: jest.fn(),
  open: false,
};

describe('ListPopper', () => {
  beforeAll(() => {
    jest.useFakeTimers();
  });
  afterAll(() => {
    jest.useRealTimers();
  });

  it('Should render correctly', () => {
    const { container } = render(<ListPopper {...props} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('Should open the list', () => {
    const { getByText } = render(<ListPopper {...props} open />);
    expect(getByText('List of lists')).toBeInTheDocument();
  });

  it('Should close the list on click outside', () => {
    const onClose = jest.fn();
    render(<ListPopper {...props} onClose={onClose} open />);
    jest.advanceTimersByTime(0);
    fireEvent.click(document.body);
    expect(onClose).toHaveBeenCalledTimes(1);
  });
});
