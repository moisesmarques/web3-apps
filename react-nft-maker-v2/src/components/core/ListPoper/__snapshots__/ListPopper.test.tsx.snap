// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListPopper Should render correctly 1`] = `
<div
  class="css-or3pe7"
>
  <div>
    <button
      type="button"
    >
      Open list
    </button>
  </div>
</div>
`;
